### PR TITLE
QLD data bits, parser refactor

### PR DESCRIPTION
**Pull request by s3364859**
@6nop @s3364942 @s3494110 @s3482247 

As the title says:
 - QLD parser
 - Some modifications of the migrations
 - Refactor of the parsers to consolidate function
 - Added in transactional bulk updates rather than multiple single inserts

bail

### CODE REVIEW

File: CPT331.Data.Migration/Migrations/M0003.cs [hunk 1]
```diff
@@ -30,8 +30,8 @@ public override void Up()
 
 			//	Offences from NSW
 			Insert.IntoTable("Offence").InSchema("Crime")
-				.Row(new { Name = "Abduction and kidnapping" })
-				.Row(new { Name = "Against justice procedures" })
+				.Row(new { Name = "Abduction and Kidnapping" })
```
Comment (s3364859) [inline CPT331.Data.Migration/Migrations/M0003.cs]: Some of these bits were incorrectly formatted upstream.

File: CPT331.Data.Migration/Migrations/M0003.cs [hunk 2]
```diff
@@ -59,6 +59,93 @@ public override void Up()
 				.Row(new { Name = "Property and deception offences" })
 				.Row(new { Name = "Public order and security offences" });
 
+			//	Offences from QLD
+			Insert.IntoTable("Offence").InSchema("Crime")
```
Comment (s3364859) [inline CPT331.Data.Migration/Migrations/M0003.cs]: Thinking of moving all this stuff to a flat model.

File: CPT331.Data.Migration/Migrations/M0006.cs [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+	{
+		public override void Down()
+		{
+			//	Nothing to see here
+		}
+
+		public override void Up()
+		{
+			//	Obviously this isn't going to work for you if you don't have the file
+			QldXmlParser qldXmlParser = new QldXmlParser(@"C:\Users\carl.belle\Downloads\Crime Data\division_Reported_Offences_Rates.xml");
```
Comment (s3364859) [inline CPT331.Data.Migration/Migrations/M0006.cs]: A very naughty hard-coded path that I will definitely fix up later (please don't hurt me @6nop).

File: CPT331.Data.Parsers/QldXmlParser.cs [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+
+using CPT331.Core.ObjectModel;
+
+#endregion
+
+namespace CPT331.Data.Parsers
+{
+	public class QldXmlParser
+	{
+		public QldXmlParser(string filename)
```
Comment (s3364859) [inline CPT331.Data.Parsers/QldXmlParser.cs]: Blah blah process my data please.

File: CPT331.Core/ObjectModel/Crime.cs [hunk 5]
```diff
@@ -8,6 +8,11 @@ namespace CPT331.Core.ObjectModel
 {
 	public class Crime
 	{
+		public Crime(int count, int localGovernmentAreaID, int month, int offenceID, int year)
```
Comment (s3364859) [inline CPT331.Core/ObjectModel/Crime.cs]: New constructor for Crime information that hasn't been stored yet.

File: CPT331.Data.Migration/App.config [hunk 6]
```diff
@@ -14,7 +14,7 @@
   <applicationSettings>
     <CPT331.Data.Migration.ApplicationConfig>
       <setting name="CrimeDataFolder" serializeAs="String">
-        <value>D:\Downloads\Crime Data</value>
+        <value>D:\Dropbox\COSC2650 Programming Project 1\Crime Data\XML Data Sources</value>
```
Comment (s3364859) [inline CPT331.Data.Migration/App.config]: Err - whoops.

File: CPT331.Data.Parsers/CPT331.Data.Parsers.csproj [hunk 7]
```diff
@@ -41,10 +41,10 @@
     <Reference Include="System.Xml" />
   </ItemGroup>
   <ItemGroup>
+    <Compile Include="QldXmlParser.cs" />
     <Compile Include="NswXmlParser.cs" />
     <Compile Include="Parser.cs" />
     <Compile Include="Properties\AssemblyInfo.cs" />
-    <Compile Include="SpreadsheetParser.cs" />
```
Comment (s3364859) [inline CPT331.Data.Parsers/CPT331.Data.Parsers.csproj]: Not required.

File: CPT331.Data.Parsers/Parser.cs [hunk 8]
```diff
@@ -16,10 +20,41 @@ public Parser(string fileName)
...
+		{
+			if ((String.IsNullOrEmpty(_fileName) == false) && (File.Exists(_fileName) == true))
+			{
+				List<Crime> crimes = new List<Crime>();
+
+				OnParse(_fileName, crimes);
+
+				if (crimes.Count > 0)
+				{
+					Commit(crimes);
```
Comment (s3364859) [inline CPT331.Data.Parsers/Parser.cs]: Yep, one of the best lines of code I've ever written.

Comment (6nop) [inline CPT331.Data.Parsers/Parser.cs]: Haha.. yes indeed.

A criminal will always do what they do best.

File: CPT331.Data.Migration/App.config (hunk 6 above)
Comment (6nop) [inline CPT331.Data.Migration/App.config]: Should be okay. 

I was thinking about using transformations (using https://visualstudiogallery.msdn.microsoft.com/579d3a78-3bdd-497c-bc21-aa6e6abbc859), but this doesn't overcome the need to check-in the transform file.

Not sure if there is a way to have user level transformations that aren't tied to the csproj file.

I have d:\, so I've just moved my dropbox folders 👍

**Comment by 6nop** [comment]
Changes look good. Hitting the merge button now.

Image of Explosion

**Pull request merged**